src/common/config.c

```diff
@@ -0,0 +1,295 @@
+#define CONFIG_PERSISTANT_MAGIC     (0xCF1C0F11)
+#define CONFIG_HEADER_SIZE          (sizeof(uint32_t) + 4) /* Magic number plus four reserved bytes */
+
+#define CONFIG_ENTRIES_START        (CONFIG_ADDRESS + CONFIG_HEADER_SIZE)
+
+uint32_t __code __at(CONFIG_ADDRESS) config_magic;
+uint8_t __code __at(CONFIG_ENTRIES_START) config_persistant[CONFIG_SIZE - CONFIG_HEADER_SIZE];
+
+#define CONFIG_FLASH_ENTRY_EMPTY            (0xFFFFFFFF)
+#define CONFIG_FLASH_ENTRY_CURRENT          (0xFF)
+#define CONFIG_FLASH_ENTRY_OBSOLETE         (0x00)
+
+// config_persistant format
+// (4) bytes: Configuration ID, 0xFFFFFFFF when an entry is not set.
+// (1) bytes: Entry current (valid) or obsolete (invalid)
+// (1) bytes: Entry type (0xFF: 32bit value follows)
+// (4) bytes: Configuration Value
+#define CONFIG_FLASH_ENTRY_ID_OFFSET        (0)
+#define CONFIG_FLASH_ENTRY_CURRENT_OFFSET   (4)
+#define CONFIG_FLASH_ENTRY_TYPE_OFFSET      (5) /* Unused - RFU */
+#define CONFIG_FLASH_ENTRY_VALUE_OFFSET     (6)
+#define CONFIG_FLASH_ENTRY_SIZE             (10)
+
+uint8_t config_validate_magic(void) {
+    uint32_t magic = flash_read_u32(CONFIG_ADDRESS);
+
+    return (CONFIG_PERSISTANT_MAGIC == magic);
+}
+
+uint32_t config_find_free_entry() {
+    if (!config_validate_magic()) {
+        // Write magic.
+        flash_write_u32(CONFIG_ADDRESS, CONFIG_PERSISTANT_MAGIC);
+
+        // Check that that magic wrote properly. If not, we'll need to erase.
+        if (!config_validate_magic()) {
+            return 0;
+        }
+    }
+
+    uint32_t addr = CONFIG_ENTRIES_START;
+    uint32_t limit = addr + sizeof(config_persistant);
+    while(addr < limit) {
+        // configuration present.
+        uint32_t id;
+        id = flash_read_u32(addr + CONFIG_FLASH_ENTRY_ID_OFFSET);
+
+        if(CONFIG_FLASH_ENTRY_EMPTY == id) {
+            // Free entry found.
+            return addr;
+        }
+
+        // Go to the next entry.
+        addr += CONFIG_FLASH_ENTRY_SIZE;
+    }
+
+    return 0;
+}
+
+void config_save_entry_id(uint32_t addr, uint32_t id) {
+    flash_write_u32(addr + CONFIG_FLASH_ENTRY_ID_OFFSET, id);
+}
+
+void config_save_entry_value(uint32_t addr, int32_t value) {
+    flash_write_u32(addr + CONFIG_FLASH_ENTRY_VALUE_OFFSET, value);
+}
+
+bool config_save_entry(config_t *entry) {
+    uint32_t id = (uint32_t)entry->config_id[3] << 24 | (uint32_t)entry->config_id[2] << 16 | (uint32_t)entry->config_id[1] << 8 | entry->config_id[0];
+    int32_t value = entry->value.value;
+    uint32_t addr = config_find_free_entry();
+
+    if (addr && ((addr + CONFIG_FLASH_ENTRY_SIZE - CONFIG_ADDRESS) < sizeof(config_persistant))) {
+        DEBUG("Saving config: %s", entry->config_short);
+        DEBUG(" = %ld\n", value);
+
+        config_save_entry_id(addr, id);
+
+        config_save_entry_value(addr, value);
+
+        return true;
+    } else {
+        return false;
+    }
+}
+
+void config_save_entries() __reentrant {
```

Comment:
What happens in the case that the config entries even when compacted overflow available flash space?

Reply:
If the compacted area overflows config space, the end result is that some of the values will not be saved.

All entries up until the config space is full will be saved. Once it's full, however, config_save_entry will return false and config_save_entries will silently fail.

Note that in this case, There's a fundamental issue in that more configuration parameters exist than are possible to store, and the only possible solutions are to either (1) reduce the number of config entries or (2) increase the config space from 1KB to larger.

I can update this code to have config_save_entries return a bool and propagate that up through to the host config_compact command. During regression testing for release, it'd then be a good idea to always issue a config_compact command and check for success to ensure that the firmware doesn't have this problem.

Reply:
The error should now be propagated to the host both when setting a config value (if compacting is needed and it fails) and if compacting is manually requested.